File: test/blas/bf16/level1/check_bscal.c
//    SPDX-License-Identifier: LGPL-3.0-or-later
/*
   This file is part of HPBLAS, a BLAS Implementation for Half-Precision
   Copyright (C) 2025 Martin Koehler

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 3 of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
   */

#include <stdio.h>
#include <stdbool.h>
#include "lpf_internal.h"
#include "bf16_helper.h"



void LPF_GLOBAL(bscal,BSCAL)(lpf_blas_int_t *, lpf_bfloat16_t *, lpf_bfloat16_t *, lpf_blas_int_t *);

// Testfunktion check_bnrm2
void check_bscal( bool *ok) {
#define NTEST 6
    lpf_blas_int_t n[NTEST]    = { 5, 3, 4, 4, 1, 0};
    lpf_blas_int_t incx[NTEST] = { 1, 1, 1, 2, 1, 1};
    lpf_bfloat16_t data[NTEST][10] =  {
        {1.0, -2.0, 3.0, -4.0, 5.0, 0,0,0,0,0},
        {1.5, -2.5, 3.5,0,0,0,0,0,0,0},
        {0,0,0,0,0,0,0,0,0,0},
        {1, 2, -1, 2, 1, 2, -1, 2, 1, 2 },
        {10,0,0,0,0,0,0,0,0,0 },
        {9, 9, 9, 9, 9 , 9, 9, 9, 9, 9}
    };
    lpf_bfloat16_t sa [ NTEST ] = { 0.1, 0.0625, 1, -4, 10, 0};
    lpf_bfloat16_t bscal_expected[NTEST][10] = {
        {0.1, -0.2, 0.3, -0.4, 0.5, 0,0,0,0,0},
        {0.09375, -0.15625, 0.21875,0,0,0,0,0,0,0},
        {0,0,0,0,0,0,0,0,0,0},
        {-4, 2, 4, 2, -4, 2, 4, 2, 1, 2 },
        {100,0,0,0,0,0,0,0,0,0 },
        {9, 9, 9, 9, 9 , 9, 9, 9, 9, 9}

    };

    for ( int i = 0; i < NTEST; i++) {
        bool lok = true;
        LPF_GLOBAL(bscal, BSCAL)(&n[i], &sa[i], data[i], &incx[i]);

        for ( int k = 0;  k < 10; k++) {
            if ( diff_bf16(data[i][k], bscal_expected[i][k]) > __FLT16_EPSILON__ * 10) {
                printf("BSCAL   -- FAIL -- Testcase %2d: N = %3d, INCX = %3d, K = %2d, Result = %10f, Expected = %10f\n", i+1, n[i], incx[i], k+1, (float)data[i][k], (float)bscal_expected[i][k]);
                *ok = false;
                lok = false;
            }
        }
        if (lok) {
            printf("BSCAL   -- PASS -- Testcase %2d: N = %3d, INCX = %3d, SA = %10f\n", i+1, n[i], incx[i], (float)sa[i]);
        }
    }
}
int main(int argc, char *argv[])
{

    (void) argc;
    (void) argv;

    bool ok = true;
    check_bscal(&ok);
    return (ok)?0:1;
}

File: test/blas/bf16/level1/check_bscal.f90
!  SPDX-License-Identifier: LGPL-3.0-or-later
PROGRAM check_bscal
    USE iso_fortran_env, only: real32, real64, int64
    USE LPF_BF16
    USE lpf_blas_bf16
    IMPLICIT NONE

    LOGICAL :: ok
    ok = .TRUE.

    CALL run_tests(ok)

    IF (.NOT. ok) THEN
        PRINT *, "HSCAL tests failed!"
        STOP 1
    END IF

CONTAINS

    SUBROUTINE run_tests(ok)
        LOGICAL, INTENT(INOUT) :: ok
        INTEGER, PARAMETER :: NTEST = 6
        INTEGER(int64), dimension(NTEST) :: n = [5, 3, 4, 4, 1, 0]
        INTEGER(int64), dimension(NTEST) :: incx = [1, 1, 1, 2, 1, 1]
        TYPE(BF16), dimension(10, NTEST) :: data
        TYPE(BF16), dimension(NTEST) :: sa
        TYPE(BF16), dimension(10, NTEST) :: expected
        INTEGER :: k, i
        LOGICAL :: lok

        ! Initialize data
        ! Testcase 1: {1.0, -2.0, 3.0, -4.0, 5.0}
        data(1, 1) = BF16(1.0_real32)
        data(2, 1) = BF16(-2.0_real32)
        data(3, 1) = BF16(3.0_real32)
        data(4, 1) = BF16(-4.0_real32)
        data(5, 1) = BF16(5.0_real32)
        data(6, 1) = BF16(0.0_real32)
        data(7, 1) = BF16(0.0_real32)
        data(8, 1) = BF16(0.0_real32)
        data(9, 1) = BF16(0.0_real32)
        data(10, 1) = BF16(0.0_real32)
        ! Testcase 2: {1.5, -2.5, 3.5}
        data(1, 2) = BF16(1.5_real32)
        data(2, 2) = BF16(-2.5_real32)
        data(3, 2) = BF16(3.5_real32)
        data(4, 2) = BF16(0.0_real32)
        data(5, 2) = BF16(0.0_real32)
        data(6, 2) = BF16(0.0_real32)
        data(7, 2) = BF16(0.0_real32)
        data(8, 2) = BF16(0.0_real32)
        data(9, 2) = BF16(0.0_real32)
        data(10, 2) = BF16(0.0_real32)
        ! Testcase 3: {0, 0, 0, 0}
        data(:, 3) = 0.0_real32
        ! Testcase 4: {1, 2, -1, 2, 1, 2, -1, 2, 1, 2}
        data(1, 4) = BF16(1.0_real32)
        data(2, 4) = BF16(2.0_real32)
        data(3, 4) = BF16(-1.0_real32)
        data(4, 4) = BF16(2.0_real32)
        data(5, 4) = BF16(1.0_real32)
        data(6, 4) = BF16(2.0_real32)
        data(7, 4) = BF16(-1.0_real32)
        data(8, 4) = BF16(2.0_real32)
        data(9, 4) = BF16(1.0_real32)
        data(10, 4) = BF16(2.0_real32)
        ! Testcase 5: {10}
        data(1, 5) = BF16(10.0_real32)
        data(2, 5) = BF16(0.0_real32)
        data(3, 5) = BF16(0.0_real32)
        data(4, 5) = BF16(0.0_real32)
        data(5, 5) = BF16(0.0_real32)
        data(6, 5) = BF16(0.0_real32)
        data(7, 5) = BF16(0.0_real32)
        data(8, 5) = BF16(0.0_real32)
        data(9, 5) = BF16(0.0_real32)
        data(10, 5) = BF16(0.0_real32)
        ! Testcase 6: {9...} but N=0
        data(:, 6) = 9.0_real32

        sa(1) = BF16(0.1_real32)
        sa(2) = BF16(0.0625_real32)
        sa(3) = BF16(1.0_real32)
        sa(4) = BF16(-4.0_real32)
        sa(5) = BF16(10.0_real32)
        sa(6) = BF16(0.0_real32)

        expected(1, 1) = BF16(0.1_real32)
        expected(2, 1) = BF16(-0.2_real32)
        expected(3, 1) = BF16(0.3_real32)
        expected(4, 1) = BF16(-0.4_real32)
        expected(5, 1) = BF16(0.5_real32)
        expected(6, 1) = BF16(0.0_real32)
        expected(7, 1) = BF16(0.0_real32)
        expected(8, 1) = BF16(0.0_real32)
        expected(9, 1) = BF16(0.0_real32)
        expected(10, 1) = BF16(0.0_real32)
        expected(1, 2) = BF16(0.09375_real32)
        expected(2, 2) = BF16(-0.15625_real32)
        expected(3, 2) = BF16(0.21875_real32)
        expected(4, 2) = BF16(0.0_real32)
        expected(5, 2) = BF16(0.0_real32)
        expected(6, 2) = BF16(0.0_real32)
        expected(7, 2) = BF16(0.0_real32)
        expected(8, 2) = BF16(0.0_real32)
        expected(9, 2) = BF16(0.0_real32)
        expected(10, 2) = BF16(0.0_real32)
        expected(:, 3) = 0.0_real32
        expected(1, 4) = BF16(-4.0_real32)
        expected(2, 4) = BF16(2.0_real32)
        expected(3, 4) = BF16(4.0_real32)
        expected(4, 4) = BF16(2.0_real32)
        expected(5, 4) = BF16(-4.0_real32)
        expected(6, 4) = BF16(2.0_real32)
        expected(7, 4) = BF16(4.0_real32)
        expected(8, 4) = BF16(2.0_real32)
        expected(9, 4) = BF16(1.0_real32)
        expected(10, 4) = BF16(2.0_real32)
        expected(1, 5) = BF16(100.0_real32)
        expected(2, 5) = BF16(0.0_real32)
        expected(3, 5) = BF16(0.0_real32)
        expected(4, 5) = BF16(0.0_real32)
        expected(5, 5) = BF16(0.0_real32)
        expected(6, 5) = BF16(0.0_real32)
        expected(7, 5) = BF16(0.0_real32)
        expected(8, 5) = BF16(0.0_real32)
        expected(9, 5) = BF16(0.0_real32)
        expected(10, 5) = BF16(0.0_real32)
        expected(1, 6) = BF16(9.0_real32)
        expected(2, 6) = BF16(9.0_real32)
        expected(3, 6) = BF16(9.0_real32)
        expected(4, 6) = BF16(9.0_real32)
        expected(5, 6) = BF16(9.0_real32)
        expected(6, 6) = BF16(9.0_real32)
        expected(7, 6) = BF16(9.0_real32)
        expected(8, 6) = BF16(9.0_real32)
        expected(9, 6) = BF16(9.0_real32)
        expected(10, 6) = BF16(9.0_real32)

        do i = 1, NTEST
            CALL scal(n(i), sa(i), data(:, i), incx(i))

            lok = .TRUE.
            do k = 1, 10
                IF (abs(dble(data(k, i)) - dble(expected(k, i))) > 1.0d-2 * max(abs(dble(expected(k, i))), 1.0d-5)) THEN
                    print "(A, I2, A, I3, A, I3, A, I2, A, F10.6, A, F10.6)", &
                          "HSCAL   -- FAIL -- Testcase ", i, ": N = ", n(i), ", INCX = ", incx(i), &
                          ", K = ", k, ", Result = ", dble(data(k, i)), &
                          ", Expected = ", dble(expected(k, i))
                    ok = .FALSE.
                    lok = .FALSE.
                END IF
            end do
            IF (lok) THEN
                print "(A, I2, A, I3, A, I3, A, F10.6)", &
                      "HSCAL   -- PASS -- Testcase ", i, ": N = ", n(i), ", INCX = ", incx(i), &
                      ", SA = ", dble(sa(i))
            END IF
        end do
    end SUBROUTINE
END PROGRAM check_bscal
